### ISSUE TITLE
Rspamd 3: conflict between MAILCOW_WHITE, SOGO_CONTACTS and user score thresholds

### ISSUE THREAD

**Issue by dragoangel** [mailcow/mailcow-dockerized#4329]
**Prior to placing the issue, please check following:** *(fill out each checkbox with an `X` once done)*
- [X] I understand that not following or deleting the below instructions will result in immediate closure and/or deletion of my issue.
- [X] I have understood that this bug report is dedicated for bugs, and not for support-related inquiries.
- [X] I have understood that answers are voluntary and community-driven, and not commercial support.
- [X] I have verified that my issue has not been already answered in the past. I also checked previous [issues](https://github.com/mailcow/mailcow-dockerized/issues).

## Summary

With Rspamd 3.1 validation check was added that throws alert when multiply `apply "default"` rules was been triggered.
Mailcow https://github.com/mailcow/mailcow-dockerized/blob/master/data/conf/rspamd/dynmaps/settings.php use this approach with `MAILCOW_WHITE`, `MAILCOW_BLACK` (whitelist|blacklist_mime_internalexampleorg && whitelist|blacklist_internalexampleorg) and could be triggered additionally if user have `MAILCOW_WHITE|BLACK` and `SOGO_CONTACT` at once - 3 times `apply "default"`.
## Logs

```
#40(normal) <8ec536>; task; lua_metric_symbol_callback: call to (SETTINGS_CHECK) failed (2): /usr/share/rspamd/plugins/settings.lua:45: invalid invocation: settings has been already set; trace: [1]:{[C]:-1 - set_settings [C]}; [2]:{/usr/share/rspamd/plugins/settings.lua:45 - apply_settings [Lua]}; [3]:{/usr/share/rspamd/plugins/settings.lua:379 - <unknown> [Lua]};
```
## Reproduction

1. logic as email user
2. go to spam tab
3. add external email to whitelist or blacklist
4. send message from external email
5. check rspamd logs

**Comment by dragoangel** [comment]
Hi @andryyy @mkuron can you check this please when have time?
Vsevolod said using multiply times `apply` is wrong and said that we need change configs, and not use this approach.
If remove check only from_mime - it will work without errors till `whitelist_mime_<?=$username_sane;?>` will not face `whitelist_sogo_<?=$username_sane;?>`. So don't see a clean solution at this point.

**Comment by dragoangel** [comment]
Hi @andryyy @mkuron can you check this please when have time?
Vsevolod said using multiply times `apply` is wrong and said that we need change configs, and not use this approach.
If leave only check for `from_mime = ...` - it will work without errors till `whitelist_mime_<?=$username_sane;?>` will not face `whitelist_sogo_<?=$username_sane;?>`. So don't see a clean solution at this point that will resolve all. Need do something with `apply "default"`.

Now if user will have `SOGO_CONTACT` and `MAILCOW_WHITE` at once - only `MAILCOW_WHITE` will appear. 
Generally it's ok, but Rspamd logic says it's wrong, because for Rspamd unclear what really should be `applied by default`.

Andre created for me `rspamd:1.80` image, and there more details trowed to logs.
Note: you need to change log level to `info` in `data/conf/rspamd/override.d/logging.inc`:
```
2021-11-18 18:21:45 #40(normal) <ee6d7c>; lua; settings.lua:48: cannot apply settings rule whitelist_mime_useremail (id=nil): settings has been already applied by rule whitelist_useremail (id=nil)
```
And to note: in Rspamd 3.2 version that build in `rspamd:1.80` email get score of `-1998` (double applied `-999`).

**Comment by mkuron** [comment]
Thanks for reporting this. The only we we can avoid this is when we make sure that one message never matches contradicting rules. I guess we can safely assume that nobody will put the same address on both the blacklist and whitelist and expect that to work. That leaves SOGO_CONTACT: it always contradicts with MAILCOW_WHITE, so we should remove it for rspamd 3 compatibility.

I guess the official way to implement this functionality with rspamd would be using multimaps. Accordig to the documentation, since rspamd 2 "dependent maps" can be used to have different maps for different users. They are not suitable for blacklisting/whitelisting because they can‘t differentiate between header-from and envelope-from, but they would work for SOGO_CONTACT where we don‘t need that.

**Comment by dragoangel** [comment]
@mkuron honestly I'd better remove MAILCOW_WHITE, as SOGo Contact is automatically handled whitelist - this one of most cool futures we have I think.

**Comment by dragoangel** [comment]
@mkuron why you don't have telegram? 😋

**Comment by mkuron** [comment]
We need the whitelist to apply to both envelope and header From. Otherwise it won‘t work with forwarding hosts. That’s why we have two separate whitelist rules A multimap will only apply to the header (unless none is present, in which case it falls back to the envelope). That is enough for SOGO_CONTACT, where we only look at the header, not the envelope.

**Comment by dragoangel** [comment]
yeah, I know... and well this is a problem. I think - can we create a script that will handle all existing rules provided by `docker-compose exec dovecot-mailcow curl http://nginx:8081/settings.php`, without `apply` syntax, save them to rspamd config directory. Run script once a minute by trigger to update this settings...

Problems;
1. limit writes to fs - if possible need to write only when existing settings are different from new
2. rspamd need to reloaded or no? - many settings rspamd take dynamically, but not all, need to check if implement

**Comment by dragoangel** [comment]
> Accordig to the documentation, since rspamd 2 "dependent maps" can be used to have different maps for different users

So we can migrate SOGo Contact rules there? Because I don't deep at that part and docs too: https://rspamd.com/doc/modules/multimap.html#dependent-maps

@mkuron you have xp with that?

Tomorrow I will try remove "default" part from apply and will play with priority. Not think it will help, but it could and deserve a try I think.

**Comment by dragoangel** [comment]
@mkuron also per user `custom score actions` settings will be always "in conflict" with all other rules. I tried removing `"default"` and leave only `apply`, but as I was thinking it wasn't change anything. Also `settings.php` already have `priorities` set, I was missed that.
I don't see a way to create real per user settings which merge all in one apply rule:

1. domain level blacklist\whitelist
2. per user blacklist\whitelist
3. actions scores
4. sogo contacts

I'm out of ideas how to handle it via settings.php dynmap.

**Comment by mkuron** [comment]
I forgot about the per-user score thresholds, because I don't use them. The conflict between MAILCOW_WHITE and SOGO_CONTACTS is actually not a big issue in practice -- just don't put your contacts on the whitelist. Similarly, don't put the same address on the blacklist and whitelist simultaneously. So in my case, everything is working correctly because I don't end up with multiple `apply`s matching on the same message.

That being said, we indeed can't keep it this way, it's just too confusing to the user. Enforcing that whitelist entries cannot match SOGo contacts or blacklist entries would be one option, but that would require dropping support for custom score threshold, which I guess many people like.

I guess a complete solution would look like this: Remove whitelist, blacklist and SOGo contacts from the dynmap. Keep the custom score thresholds there. Migrate all of the address lists to "dependent multimaps", though I don't think it's documented how they work. Reading the documentation more closely, I think we can continue matching on `From` headers and envelope `FROM`s separately:
```
mailcow_env_white {
  type = "from";
  filter = "email";
  map = [
    "redis+selector://id('MAILCOW_WHITE_');to:addr",
    "redis+selector://id('MAILCOW_WHITE_');to:domain"
  ];
  symbol = "MAILCOW_WHITE";
  description = "Envelope From is whitelisted";
  score = -999.0;
}

mailcow_white {
  type = "header";
  header = "from";
  filter = "email";
  map = [
    "redis+selector://id('MAILCOW_WHITE_');to:addr",
    "redis+selector://id('MAILCOW_WHITE_');to:domain"
  ];
  symbol = "MAILCOW_WHITE";
  description = "Header From is whitelisted";
  score = -999.0;
}

mailcow_env_black {
  type = "from";
  filter = "email";
  map = [
    "redis+selector://id('MAILCOW_BLACK_');to:addr",
    "redis+selector://id('MAILCOW_BLACK_');to:domain"
  ];
  symbol = "MAILCOW_BLACK";
  description = "Envelope From is blacklisted";
  score = 999.0;
}

mailcow_black {
  type = "header";
  header = "from";
  filter = "email";
  map = [
    "redis+selector://id('MAILCOW_BLACK_');to:addr",
    "redis+selector://id('MAILCOW_BLACK_');to:domain"
  ];
  symbol = "MAILCOW_BLACK";
  description = "Header From is blacklisted";
  score = 999.0;
}

sogo_contact {
  type = "header";
  type = "from";
  filter = "email";
  map = [
    "redis+selector://id('SOGO_CONTACT_');to:addr"
  ];
  symbol = "SOGO_CONTACT";
  description = "Header From is in SOGo contacts";
  score = -99.0;
}
```
Of course, we need to get all that stuff into redis now. And if one message is sent to multiple recipients on the same Mailcow at the same time, I'm not sure whose whitelist/blacklist/contacts it will match.

**Comment by dragoangel** [comment]
I can help with testing if needed of rspamd to check how it will go in any cases, but my knowledge of php not good enough to help.

**Comment by mkuron** [comment]
This is definitely a bug and tagged as such, so I'm not sure why the bot tagged it as stale -- it's not.

**Comment by MAGICCC** [comment]
Because the bot doesnt exempt issues or PRs when the bug label is set.

**Comment by MAGICCC** [comment]
I added a new label 'neverstale' to avoid closing important issues/PRs
